File: android/src/main/kotlin/com/emarsys/emarsys_sdk/command/EmarsysCommandFactory.kt
package com.emarsys.emarsys_sdk.command

import android.app.Application
import android.content.SharedPreferences
import android.os.Handler
import com.emarsys.emarsys_sdk.command.config.*
import com.emarsys.emarsys_sdk.command.geofence.*
import com.emarsys.emarsys_sdk.command.inapp.InAppIsPausedCommand
import com.emarsys.emarsys_sdk.command.inapp.InAppPauseCommand
import com.emarsys.emarsys_sdk.command.inapp.InAppResumeCommand
import com.emarsys.emarsys_sdk.command.mobileengage.TrackCustomEventCommand
import com.emarsys.emarsys_sdk.command.mobileengage.contact.ClearContactCommand
import com.emarsys.emarsys_sdk.command.mobileengage.contact.SetContactCommand
import com.emarsys.emarsys_sdk.command.mobileengage.inbox.AddTagCommand
import com.emarsys.emarsys_sdk.command.mobileengage.inbox.FetchMessagesCommand
import com.emarsys.emarsys_sdk.command.mobileengage.inbox.RemoveTagCommand
import com.emarsys.emarsys_sdk.command.mobileengage.push.PushSendingEnabledCommand
import com.emarsys.emarsys_sdk.command.mobileengage.push.RegisterNotificationChannelsCommand
import com.emarsys.emarsys_sdk.command.predict.*
import com.emarsys.emarsys_sdk.command.setup.InitializeCommand
import com.emarsys.emarsys_sdk.command.setup.SetupCommand
import com.emarsys.emarsys_sdk.mapper.*
import com.emarsys.emarsys_sdk.notification.NotificationChannelFactory
import com.emarsys.emarsys_sdk.storage.PushTokenStorage


class EmarsysCommandFactory(
        private val application: Application,
        private val pushTokenStorage: PushTokenStorage,
        private val setupCacheSharedPreferences: SharedPreferences,
        private val flutterWrapperSharedPreferences: SharedPreferences,
        private val notificationChannelFactory: NotificationChannelFactory,
        private val inboxResultMapper: InboxResultMapper,
        private val backgroundHandler: Handler,
        private val geofenceMapper: GeofenceMapper,
        private val mapToProductMapper: MapToProductMapper,
        private val recommendationLogicMapper: RecommendationLogicMapper,
        private val recommendationFilterListMapper: RecommendationFilterListMapper,
        private val productMapper: ProductMapper
) {

    fun create(methodName: String): EmarsysCommand? {
        return when (methodName) {
            "setup" -> SetupCommand(
                    application, pushTokenStorage,
                    setupCacheSharedPreferences, false
            )
            "android.setupFromCache" -> SetupCommand(
                    application, pushTokenStorage,
                    setupCacheSharedPreferences, true
            )
            "setContact" -> SetContactCommand()
            "clearContact" -> ClearContactCommand()
            "android.initialize" -> InitializeCommand(
                    flutterWrapperSharedPreferences,
                    backgroundHandler
            )
            "push.pushSendingEnabled" -> PushSendingEnabledCommand(pushTokenStorage)
            "push.android.registerNotificationChannels" -> RegisterNotificationChannelsCommand(
                    application,
                    notificationChannelFactory
            )
            "config.changeApplicationCode" -> ChangeApplicationCodeCommand(
                    setupCacheSharedPreferences
            )
            "config.changeMerchantId" -> ChangeMerchantIdCommand(setupCacheSharedPreferences)
            "config.applicationCode" -> ApplicationCodeCommand()
            "config.merchantId" -> MerchantIdCommand()
            "config.contactFieldId" -> ContactFieldIdCommand()
            "config.hardwareId" -> HardwareIdCommand()
            "config.languageCode" -> LanguageCodeCommand()
            "config.notificationSettings" -> NotificationSettingsCommand()
            "config.sdkVersion" -> SdkVersionCommand()
            "trackCustomEvent" -> TrackCustomEventCommand()
            "inbox.fetchMessages" -> FetchMessagesCommand(inboxResultMapper)
            "inbox.addTag" -> AddTagCommand()
            "inbox.removeTag" -> RemoveTagCommand()
            "geofence.enable" -> GeofenceEnableCommand()
            "geofence.disable" -> GeofenceDisableCommand()
            "geofence.initialEnterTriggerEnabled" -> GeofenceInitialEnterTriggerEnabledCommand()
            "geofence.isEnabled" -> GeofenceIsEnabledCommand()
            "geofence.registeredGeofences" -> RegisteredGeofencesCommand(geofenceMapper)
            "inApp.resume" -> InAppResumeCommand()
            "inApp.pause" -> InAppPauseCommand()
            "inApp.isPaused" -> InAppIsPausedCommand()
            "predict.trackItemView" -> TrackItemViewCommand()
            "predict.trackCategoryView" -> TrackCategoryViewCommand()
            "predict.trackTag" -> TrackTagCommand()
            "predict.trackSearchTerm" -> TrackSearchTermCommand()
            "predict.trackCart" -> TrackCartItemCommand()
            "predict.trackPurchase" -> TrackPurchaseCommand()
            "predict.trackRecommendationClick" -> TrackRecommendationClickCommand(mapToProductMapper)
            "predict.recommendProducts" -> RecommendProductsCommand(
                    logicMapper = recommendationLogicMapper,
                    filterListMapper = recommendationFilterListMapper,
                    productMapper = productMapper
            )
            else -> null
        }
    }
}

File: ios/Classes/Wrapper/Commands/EmarsysCommandFactory.swift
//
// Created by Emarsys on 2021. 04. 22..
//

import EmarsysSDK

class EmarsysCommandFactory {

    var pushEventHandler: EMSEventHandlerBlock
    var silentPushEventHandler: EMSEventHandlerBlock
    var geofenceEventHandler: EMSEventHandlerBlock
    var inboxMapper: InboxMapper
    var inAppEventHandler: EMSEventHandlerBlock
    var mapToProductMapper: MapToProductMapper
    var productsMapper: ProductsMapper
    var logicMapper: LogicMapper
    var recommendationFilterMapper: RecommendationFilterMapper

    init(pushEventHandler: @escaping EMSEventHandlerBlock,
         silentPushEventHandler: @escaping EMSEventHandlerBlock,
         inboxMapper: InboxMapper,
         geofenceEventHandler: @escaping EMSEventHandlerBlock,
         inAppEventHandler: @escaping EMSEventHandlerBlock,
         mapToProductMapper: MapToProductMapper,
         productsMapper: ProductsMapper,
         logicMapper: LogicMapper,
         recommendationFilterMapper: RecommendationFilterMapper) {
        self.pushEventHandler = pushEventHandler
        self.silentPushEventHandler = silentPushEventHandler
        self.inboxMapper = inboxMapper
        self.geofenceEventHandler = geofenceEventHandler
        self.inAppEventHandler = inAppEventHandler
        self.mapToProductMapper = mapToProductMapper
        self.productsMapper = productsMapper
        self.logicMapper = logicMapper
        self.recommendationFilterMapper = recommendationFilterMapper
    }

    func create(name: String) -> EmarsysCommandProtocol? {
        var result: EmarsysCommandProtocol?
        switch name {
        case "setup":
            result = SetupCommand(pushEventHandler: self.pushEventHandler,
                    silentPushEventHandler: self.silentPushEventHandler,
                    geofenceEventHandler: self.geofenceEventHandler,
                    inAppEventHandler: self.inAppEventHandler)
        case "setContact":
            result = SetContactCommand()
        case "clearContact":
            result = ClearContactCommand()
        case "push.pushSendingEnabled":
            result = PushSendingEnabledCommand()
        case "push.android.registerNotificationChannels":
            result = EmptyCommand()
        case "config.applicationCode":
            result = ApplicationCodeCommand()
        case "config.merchantId":
            result = MerchantIdCommand()
        case "config.contactFieldId":
            result = ContactFieldIdCommand()
        case "config.hardwareId":
            result = HardwareIdCommand()
        case "config.languageCode":
            result = LanguageCodeCommand()
        case "config.notificationSettings":
            result = NotificationSettingsCommand()
        case "config.sdkVersion":
            result = SdkVersionCommand()
        case "config.changeApplicationCode":
            result = ChangeApplicationCodeCommand()
        case "config.changeMerchantId":
            result = ChangeMerchantIdCommand()
        case "trackCustomEvent":
            result = TrackCustomEventCommand()
        case "inbox.fetchMessages":
            result = FetchMessagesCommand(inboxMapper: inboxMapper)
        case "inbox.addTag":
            result = AddTagCommand()
        case "inbox.removeTag":
            result = RemoveTagCommand()
        case "geofence.enable":
            result = GeofenceEnableCommand()
        case "geofence.disable":
            result = GeofenceDisableCommand()
        case "geofence.setInitialEnterTriggerEnabled":
            result = GeofenceSetInitialEnterTriggerEnabledCommand()
        case "geofence.ios.requestAlwaysAuthorization":
            result = GeofenceiOSRequestAlwaysAuthorizationCommand()
        case "geofence.isEnabled":
            result = GeofenceisEnabledCommand()
        case "inApp.pause":
            result = InAppPauseCommand()
        case "inApp.resume":
            result = InAppResumeCommand()
        case "inApp.isPaused":
            result = InAppIsPausedCommand()
        case "predict.trackCart":
            result = TrackCartItemCommand()
        case "predict.trackPurchase":
            result = TrackPurchaseCommand()
        case "predict.trackCategoryView":
            result = TrackCategoryCommand()
        case "predict.trackSearchTerm":
            result = TrackSearchTermCommand()
        case "predict.trackTag":
            result = TrackTagCommand()
        case "predict.trackItemView":
            result = TrackItemViewCommand()
        case "predict.recommendProducts":
            result = RecommendProductsCommand(productsMapper: productsMapper, logicMapper: logicMapper, recommendationFilterMapper: recommendationFilterMapper)
        case "predict.trackRecommendationClick":
            result = TrackRecommendationClickCommand(mapToProductMapper: mapToProductMapper)
        default:
            result = nil
        }
        return result
    }

}
